src/lando/main/scm/hg.py

```diff
@@ -0,0 +1,511 @@
+    """Exception when pulling changes from the upstream repo fails."""
+
+    SNIPPETS = [
+        "abort: HTTP Error 500:",
+        "abort: push failed on remote",
+    ]
+
+
+class HgPatchConflict(PatchConflict, HgException):
+    """Exception when patch fails to apply due to a conflict."""
+
+    # TODO: Parse affected files from hg output and present
+    # them in a structured way.
+
+    SNIPPETS = [
+        "unresolved conflicts (see hg resolve",
+        "hunk FAILED -- saving rejects to file",
+        "hunks FAILED -- saving rejects to file",
+    ]
+
+
+class HgScm(AbstractScm):
+    ENCODING = "utf-8"
+    DEFAULT_CONFIGS = {
+        "ui.username": "Otto Länd <bind-autoland@mozilla.com>",
+        "ui.interactive": "False",
+        "ui.merge": "internal:merge",
+        "ui.ssh": (
+            "ssh "
+            f'-o "SendEnv {REQUEST_USER_ENV_VAR}" '
+            '-o "StrictHostKeyChecking no" '
+            '-o "PasswordAuthentication no" '
+            f'-o "User {settings.LANDING_WORKER_USERNAME}" '
+            f'-o "Port {settings.LANDING_WORKER_TARGET_SSH_PORT}"'
+        ),
+        "extensions.purge": "",
+        "extensions.strip": "",
+        "extensions.rebase": "",
+        "extensions.set_landing_system": settings.BASE_DIR
+        / "api/legacy/hgext/set_landing_system.py",
+    }
+
+    config: dict
+
+    hg_repo: hglib.client.hgclient
+
+    def __init__(self, path: str, config: Optional[dict] = None):
+        self.config = copy.copy(self.DEFAULT_CONFIGS)
+
+        # Somewhere to store patch headers for testing.
+        self.patch_header = None
+
+        if config:
+            self.config.update(config)
+
+        super().__init__(path)
+
+    @property
+    def REJECTS_PATH(self) -> Path:
+        return Path("/tmp/patch_rejects")
+
+    def push(
+        self, push_path: str, target: Optional[str] = None, force_push: bool = False
+    ) -> None:
+        bookmark = target
+        target = push_path
+        if not os.getenv(REQUEST_USER_ENV_VAR):
+            raise ValueError(f"{REQUEST_USER_ENV_VAR} not set while attempting to push")
+
+        extra_args = []
+
+        if force_push:
+            extra_args.append("-f")
+
+        if bookmark is None:
+            self.run_hg(["push", "-r", "tip", target] + extra_args)
+        else:
+            self.run_hg_cmds(
+                [
+                    ["bookmark", bookmark],
+                    ["push", "-B", bookmark, target] + extra_args,
+                ]
+            )
+
+    def last_commit_for_path(self, repo_path: str, path: str) -> str:
+        return self.run_hg(
+            [
+                "log",
+                "--cwd",
+                repo_path,
+                "--template",
+                "{node}",
+                "-l",
+                "1",
+                path,
+            ]
+        ).decode()
+
+    def apply_patch(
+        self, diff: str, commit_description: str, commit_author: str, commit_date: str
+    ):
```

Comment:
In theory, yes, but given the number of other changes around it, I think extracting it would create more conflicts that would need to be solved on the way out, and back on the way in when on the following commits, so I'm not sure it's worth it, as we'd end up in the same state.